File: ios-rockscout/RockScout/Services/ErrorReporter.swift
import Foundation
import CryptoKit

/// Central error reporter that ships runtime exceptions to the Supabase
/// `rockscout_error_logs` table via the `/error-report` Cloudflare Worker.
///
/// Self-healing: for known recoverable error patterns, the reporter can
/// automatically apply a fix before the error propagates. See `SelfHealer`.
///
/// All network calls are fire-and-forget on a background task — they never
/// block the calling thread or crash the app if the upload fails.
@MainActor
final class ErrorReporter {
    static let shared = ErrorReporter()
    private init() {}

    // MARK: - State

    private var appVersion: String = "unknown"
    private var osVersion: String = "unknown"
    private var deviceModel: String = "unknown"
    private var currentUserId: String?
    private var lastBreadcrumb: String?
    private var lastReportTime: [String: Date] = [:]
    private let dedupWindow: TimeInterval = 300 // 5 minutes

    // MARK: - Init

    func configure() {
        appVersion = Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String ?? "unknown"
        osVersion = ProcessInfo.processInfo.operatingSystemVersionString
        deviceModel = Self.deviceModel()
    }

    func setUserId(_ userId: String?) {
        currentUserId = userId
    }

    func setBreadcrumb(_ action: String) {
        lastBreadcrumb = String(action.prefix(500))
    }

    // MARK: - Report

    /// Report an error to the central service. Fire-and-forget.
    func report(
        screen: String,
        error: Error,
        isFatal: Bool = false,
        attemptSelfHeal: Bool = true
    ) {
        let errorType = String(describing: type(of: error))
        let message = error.localizedDescription
        let stackTrace = error.customMirror.children
            .compactMap { $0.label == "stackTrace" ? $0.value as? String : nil }
            .first

        var healAction: String?
        var autoHealed = false

        if attemptSelfHeal && !isFatal {
            if let result = SelfHealer.attemptHeal(error: error, screen: screen) {
                autoHealed = true
                healAction = result
            }
        }

        let fingerprint = Self.fingerprint(type: errorType, message: message, screen: screen)

        // Dedup: skip if same fingerprint was reported within the window
        if let lastTime = lastReportTime[fingerprint],
           Date().timeIntervalSince(lastTime) < dedupWindow {
            return
        }
        lastReportTime[fingerprint] = Date()

        let payload = ErrorPayload(
            platform: "ios",
            appVersion: appVersion,
            osVersion: osVersion,
            deviceModel: deviceModel,
            userId: currentUserId,
            errorType: String(errorType.prefix(300)),
            errorMessage: String(message.prefix(2000)),
            stackTrace: stackTrace.map { String($0.prefix(8000)) },
            isFatal: isFatal,
            screen: String(screen.prefix(200)),
            breadcrumb: lastBreadcrumb,
            autoHealed: autoHealed,
            healAction: healAction,
            fingerprint: fingerprint
        )

        Task.detached { [payload] in
            await Self.upload(payload)
        }
    }

    /// Report a plain message (no throwable).
    func reportMessage(
        screen: String,
        message: String,
        isFatal: Bool = false
    ) {
        let fingerprint = Self.fingerprint(type: "Message", message: message, screen: screen)

        if let lastTime = lastReportTime[fingerprint],
           Date().timeIntervalSince(lastTime) < dedupWindow {
            return
        }
        lastReportTime[fingerprint] = Date()

        let payload = ErrorPayload(
            platform: "ios",
            appVersion: appVersion,
            osVersion: osVersion,
            deviceModel: deviceModel,
            userId: currentUserId,
            errorType: "Message",
            errorMessage: String(message.prefix(2000)),
            stackTrace: nil,
            isFatal: isFatal,
            screen: String(screen.prefix(200)),
            breadcrumb: lastBreadcrumb,
            autoHealed: false,
            healAction: nil,
            fingerprint: fingerprint
        )

        Task.detached { [payload] in
            await Self.upload(payload)
        }
    }

    // MARK: - Upload

    nonisolated private static func upload(_ payload: ErrorPayload) async {
        let url = AppSecrets.functionsURL
        let appKey = AppSecrets.appKey
        guard !url.isEmpty, !appKey.isEmpty else { return }

        guard let endpoint = URL(string: "\(url)/error-report") else { return }

        var request = URLRequest(url: endpoint)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        request.setValue(appKey, forHTTPHeaderField: "X-App-Key")

        do {
            request.httpBody = try JSONEncoder().encode(payload)
        } catch {
            return
        }

        do {
            let (_, response) = try await URLSession.shared.data(for: request)
            if let http = response as? HTTPURLResponse {
                #if DEBUG
                print("[ErrorReporter] Upload status: \(http.statusCode)")
                #endif
            }
        } catch {
            // Fire-and-forget — network failures are expected
        }
    }

    // MARK: - Fingerprint

    nonisolated private static func fingerprint(type: String, message: String, screen: String) -> String {
        let input = "\(type)|\(message)|\(screen)"
        let hash = SHA256.hash(data: Data(input.utf8))
        return hash.compactMap { String(format: "%02x", $0) }.joined().prefix(16).description
    }

    // MARK: - Device

    nonisolated private static func deviceModel() -> String {
        var systemInfo = utsname()
        uname(&systemInfo)
        let machine = withUnsafePointer(to: &systemInfo.machine) { ptr in
            ptr.withMemoryRebound(to: CChar.self, capacity: 1) {
                String(cString: $0)
            }
        }
        return machine
    }
}

// MARK: - Payload

nonisolated struct ErrorPayload: Codable {
    let platform: String
    let appVersion: String
    let osVersion: String
    let deviceModel: String
    let userId: String?
    let errorType: String
    let errorMessage: String
    let stackTrace: String?
    let isFatal: Bool
    let screen: String
    let breadcrumb: String?
    let autoHealed: Bool
    let healAction: String?
    let fingerprint: String

    enum CodingKeys: String, CodingKey {
        case platform
        case appVersion = "appVersion"
        case osVersion = "osVersion"
        case deviceModel = "deviceModel"
        case userId
        case errorType
        case errorMessage
        case stackTrace
        case isFatal
        case screen
        case breadcrumb
        case autoHealed
        case healAction
        case fingerprint
    }
}

File: web/src/lib/errorReporter.ts
/**
 * Central error reporter for the web PWA.
 *
 * Ships runtime exceptions to the Supabase `rockscout_error_logs` table via
 * the `/error-report` Cloudflare Worker endpoint.
 *
 * Self-healing: for known recoverable patterns (stale cache, service worker
 * issues, missing config), the reporter automatically applies a fix before
 * the error propagates to the user.
 *
 * All network calls are fire-and-forget — they never block rendering or
 * crash the app if the upload fails.
 */

import { FUNCTIONS_URL, APP_KEY } from "@/lib/config";

interface ErrorPayload {
  platform: string;
  appVersion?: string;
  osVersion?: string;
  deviceModel?: string;
  userId?: string;
  errorType: string;
  errorMessage: string;
  stackTrace?: string;
  isFatal: boolean;
  screen?: string;
  breadcrumb?: string;
  autoHealed: boolean;
  healAction?: string;
  fingerprint: string;
}

/** Dedup window — skip if same fingerprint was reported within this time. */
const DEDUP_WINDOW_MS = 5 * 60 * 1000; // 5 minutes
const MAX_STACK_TRACE = 8000;
const MAX_SCREEN = 200;
const MAX_MESSAGE = 2000;
const MAX_BREADCRUMB = 500;

/** In-memory dedup map: fingerprint → last report timestamp. */
const reportTimestamps = new Map<string, number>();

/** Current user ID (set after auth). */
let currentUserId: string | null = null;

/** Last user action for breadcrumb tracing. */
let lastBreadcrumb: string | null = null;

/** App version from package.json or build. */
const APP_VERSION = "1.0.0";

/** Set the current user ID so errors can be attributed. */
export function setErrorReporterUserId(userId: string | null): void {
  currentUserId = userId;
}

/** Set the last user action for breadcrumb tracing. */
export function setBreadcrumb(action: string): void {
  lastBreadcrumb = action.slice(0, MAX_BREADCRUMB);
}

/** SHA-256 of (errorType + message + screen), first 16 hex chars. */
async function fingerprint(
  type: string,
  message: string,
  screen: string,
): Promise<string> {
  const input = `${type}|${message}|${screen}`;
  if (globalThis.crypto?.subtle) {
    const data = new TextEncoder().encode(input);
    const hash = await globalThis.crypto.subtle.digest("SHA-256", data);
    return Array.from(new Uint8Array(hash))
      .map((b) => b.toString(16).padStart(2, "0"))
      .join("")
      .slice(0, 16);
  }
  // Fallback for very old browsers
  let hash = 0;
  for (let i = 0; i < input.length; i++) {
    hash = ((hash << 5) - hash + input.charCodeAt(i)) | 0;
  }
  return Math.abs(hash).toString(16).padStart(16, "0").slice(0, 16);
}

/**
 * Report an error to the central service. Fire-and-forget.
 *
 * @param screen - Route/screen where the error occurred
 * @param error - The exception
 * @param isFatal - Whether this is a fatal crash
 * @param attemptSelfHeal - If true, try to auto-fix known patterns
 */
export async function reportError(
  screen: string,
  error: unknown,
  isFatal = false,
  attemptSelfHeal = true,
): Promise<void> {
  const errorObj = error instanceof Error ? error : new Error(String(error));
  const errorType = errorObj.name || "Error";
  const message = (errorObj.message || "(no message)").slice(0, MAX_MESSAGE);
  const stackTrace = errorObj.stack?.slice(0, MAX_STACK_TRACE);

  let healAction: string | null = null;
  let autoHealed = false;

  if (attemptSelfHeal && !isFatal) {
    const result = selfHeal(errorType, message, screen);
    if (result) {
      autoHealed = true;
      healAction = result;
      console.info(`[ErrorReporter] Auto-healed: ${healAction}`);
    }
  }

  const fp = await fingerprint(errorType, message, screen);

  // Dedup
  const now = Date.now();
  const lastTime = reportTimestamps.get(fp);
  if (lastTime && now - lastTime < DEDUP_WINDOW_MS) {
    return; // Skip duplicate
  }
  reportTimestamps.set(fp, now);

  const payload: ErrorPayload = {
    platform: "web",
    appVersion: APP_VERSION,
    osVersion: navigator.userAgent.includes("Mac") ? "macOS" :
      navigator.userAgent.includes("Win") ? "Windows" :
      navigator.userAgent.includes("Android") ? "Android" :
      navigator.userAgent.includes("iPhone") || navigator.userAgent.includes("iPad") ? "iOS" : "Unknown",
    deviceModel: navigator.userAgent.slice(0, 300),
    userId: currentUserId ?? undefined,
    errorType: errorType.slice(0, 300),
    errorMessage: message,
    stackTrace: stackTrace || undefined,
    isFatal,
    screen: screen.slice(0, MAX_SCREEN) || undefined,
    breadcrumb: lastBreadcrumb ?? undefined,
    autoHealed,
    healAction: healAction ?? undefined,
    fingerprint: fp,
  };

  // Fire-and-forget upload
  void uploadError(payload);
}

/**
 * Report a plain message (no exception object).
 */
export async function reportMessage(
  screen: string,
  message: string,
  isFatal = false,
): Promise<void> {
  const fp = await fingerprint("Message", message, screen);

  const now = Date.now();
  const lastTime = reportTimestamps.get(fp);
  if (lastTime && now - lastTime < DEDUP_WINDOW_MS) return;
  reportTimestamps.set(fp, now);

  const payload: ErrorPayload = {
    platform: "web",
    appVersion: APP_VERSION,
    osVersion: undefined,
    deviceModel: navigator.userAgent.slice(0, 300),
    userId: currentUserId ?? undefined,
    errorType: "Message",
    errorMessage: message.slice(0, MAX_MESSAGE),
    stackTrace: undefined,
    isFatal,
    screen: screen.slice(0, MAX_SCREEN) || undefined,
    breadcrumb: lastBreadcrumb ?? undefined,
    autoHealed: false,
    healAction: undefined,
    fingerprint: fp,
  };

  void uploadError(payload);
}

/** Upload the error payload to the central service. */
async function uploadError(payload: ErrorPayload): Promise<void> {
  if (!FUNCTIONS_URL || !APP_KEY) {
    // No backend configured — log to console as fallback
    console.error("[ErrorReporter] No backend configured:", payload.errorType, payload.errorMessage);
    return;
  }

  try {
    await fetch(`${FUNCTIONS_URL}/error-report`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        "X-App-Key": APP_KEY,
      },
      body: JSON.stringify(payload),
    });
  } catch {
    // Fire-and-forget — network failures are expected
  }
}

/**
 * Self-healing for known web error patterns.
 * Returns the heal action description if healed, or null if not recognized.
 */
function selfHeal(errorType: string, message: string, screen: string): string | null {
  const msg = message.toLowerCase();

  // 1. Chunk loading errors — stale service worker serving old chunks
  if (errorType === "SyntaxError" || msg.includes("unexpected token") ||
      msg.includes("failed to fetch dynamic import") ||
      msg.includes("loading chunk") || msg.includes("loading module")) {
    clearStaleCache();
    return "cleared_stale_service_worker";
  }

  // 2. Network / fetch errors — retry will happen automatically, just acknowledge
  if (errorType === "TypeError" && msg.includes("failed to fetch")) {
    return "network_retry_acknowledged";
  }

  // 3. Storage quota exceeded — clear non-essential caches
  if (msg.includes("quota") || msg.includes("storage") || msg.includes("exceeded")) {
    clearStorageQuota();
    return "cleared_storage_quota";
  }

  // 4. Service worker registration failure
  if (msg.includes("service worker") || msg.includes("serviceworker")) {
    // Unregister stale workers so a fresh one can register
    if ("serviceWorker" in navigator) {
      navigator.serviceWorker.getRegistrations().then((regs) => {
        regs.forEach((r) => r.unregister().catch(() => {}));
      });
    }
    return "unregistered_stale_service_worker";
  }

  // 5. React context / undefined destructuring — the ErrorBoundary already
  //    handles this with a reload button, so just acknowledge
  if (msg.includes("cannot destructure") || msg.includes("is undefined") ||
      msg.includes("is null")) {
    return null; // Let the ErrorBoundary handle it
  }

  return null;
}

/** Clear stale caches and unregister service workers. */
function clearStaleCache(): void {
  if ("caches" in window) {
    caches.keys().then((keys) =>
      Promise.all(keys.map((k) => caches.delete(k))).catch(() => {}),
    );
  }
}

/** Clear non-essential storage to free up quota. */
function clearStorageQuota(): void {
  // Clear sessionStorage (non-persistent)
  try {
    sessionStorage.clear();
  } catch {
    // Ignore
  }

  // Clear caches API
  if ("caches" in window) {
    caches.keys().then((keys) =>
      Promise.all(keys.map((k) => caches.delete(k))).catch(() => {}),
    );
  }

  // Don't clear localStorage — it has auth tokens and user prefs
  // Don't clear IndexedDB — it has Supabase state
}

/**
 * Install global error handlers for uncaught exceptions and unhandled
 * promise rejections. Call this once at app startup.
 */
export function installGlobalErrorHandlers(): void {
  // Uncaught errors (outside React error boundary)
  window.addEventListener("error", (event) => {
    void reportError(
      window.location.pathname,
      event.error ?? new Error(event.message),
      false, // isFatal
    );
  });

  // Unhandled promise rejections
  window.addEventListener("unhandledrejection", (event) => {
    const error = event.reason instanceof Error
      ? event.reason
      : new Error(String(event.reason));
    void reportError(
      window.location.pathname,
      error,
      false, // isFatal
    );
  });
}
